src/backend/distributed/commands/foreign_constraint.c

```diff
@@ -696,89 +649,125 @@ HeapTupleOfForeignConstraintIncludesColumn(HeapTuple heapTuple, Oid relationId,
 	}
 
-	systable_endscan(scanDescriptor);
-	heap_close(pgConstraint, NoLock);
-
 	return false;
 }
 
 
 /*
- * ConstraintIsAForeignKey returns true if the given constraint name
- * is a foreign key to defined on the relation.
+ * GetTableForeignConstraintOidsInternal takes in a relationId, and returns
+ * the list of OIDs for foreign constraints that the relation with relationId
+ * is involved according to "attributeNumber" argument. In that sense,
+ * "attributeNumber" should be:
+ *  - "Anum_pg_constraint_conrelid" to get the ones in which our relation is
+ *    the "referencing" one, and
+ *  - "Anum_pg_constraint_confrelid" to get the ones in which our relation is
+ *    the "referenced" one.
+ * Note that since there is no index for the column Anum_pg_constraint_confrelid,
+ * this function performs sequential scan on pg_constraint in that case.
+ *
+ * Self-referencing foreign keys will not be considered if excludeSelfReference
+ * is true.
  */
-bool
-ConstraintIsAForeignKey(char *constraintNameInput, Oid relationId)
+static List *
+GetTableForeignConstraintOidsInternal(Oid relationId, AttrNumber attributeNumber,
+									  bool excludeSelfReference, bool findAll)
 {
+	Assert(attributeNumber == Anum_pg_constraint_conrelid ||
+		   attributeNumber == Anum_pg_constraint_confrelid);
+
+	List *tableForeignConstraintOids = NIL;
+
 	ScanKeyData scanKey[1];
 	int scanKeyCount = 1;
+	bool indexOk = (attributeNumber == Anum_pg_constraint_conrelid) ? true : false;
 
 	Relation pgConstraint = heap_open(ConstraintRelationId, AccessShareLock);
-
-	ScanKeyInit(&scanKey[0], Anum_pg_constraint_contype, BTEqualStrategyNumber, F_CHAREQ,
-				CharGetDatum(CONSTRAINT_FOREIGN));
-	SysScanDesc scanDescriptor = systable_beginscan(pgConstraint, InvalidOid, false,
-													NULL, scanKeyCount, scanKey);
+	ScanKeyInit(&scanKey[0], attributeNumber, BTEqualStrategyNumber, F_OIDEQ,
+				relationId);
+	SysScanDesc scanDescriptor = systable_beginscan(pgConstraint,
+													ConstraintRelidTypidNameIndexId,
+													indexOk, NULL, scanKeyCount,
+													scanKey);
 
 	HeapTuple heapTuple = systable_getnext(scanDescriptor);
 	while (HeapTupleIsValid(heapTuple))
 	{
 		Form_pg_constraint constraintForm = (Form_pg_constraint) GETSTRUCT(heapTuple);
-		char *constraintName = (constraintForm->conname).data;
 
-		if (strncmp(constraintName, constraintNameInput, NAMEDATALEN) == 0 &&
-			constraintForm->conrelid == relationId)
+		if (constraintForm->contype != CONSTRAINT_FOREIGN)
 		{
-			systable_endscan(scanDescriptor);
-			heap_close(pgConstraint, AccessShareLock);
+			heapTuple = systable_getnext(scanDescriptor);
+			continue;
+		}
 
-			return true;
+		bool inheritedConstraint = OidIsValid(constraintForm->conparentid);
+		if (inheritedConstraint)
+		{
+			heapTuple = systable_getnext(scanDescriptor);
+			continue;
+		}
+
+		Oid constraintId = get_relation_constraint_oid_compat(heapTuple);
+
+		if (excludeSelfReference && (constraintForm->conrelid ==
```

Comment:
it could increase readability to put this into a variable like:

```
bool isSelfReferencing = constraintForm->conrelid == constraintForm->confrelid ;
```